crates/bevy_reflect/bevy_reflect_derive/src/impls/enums.rs

```diff
@@ -262,126 +261,91 @@ struct EnumImpls {
-                let mut field_info = Vec::new();
-                let mut field_idx: usize = 0;
-                for field in fields.iter() {
+                let (field_len, field_info) = for_fields!(fields, |reflect_idx, _, field| {
                     let field_ident = field.data.ident.as_ref().unwrap();
-
-                    if field.attrs.ignore {
-                        // Ignored field
-                        continue;
-                    }
-
                     let field_name = field_ident.to_string();
                     enum_field.push(quote! {
                         #unit{ #field_ident, .. } if #ref_name == #field_name => Some(#field_ident)
                     });
                     enum_field_at.push(quote! {
-                        #unit{ #field_ident, .. } if #ref_index == #field_idx => Some(#field_ident)
+                        #unit{ #field_ident, .. } if #ref_index == #reflect_idx => Some(#field_ident)
                     });
                     enum_index_of.push(quote! {
-                        #unit{ .. } if #ref_name == #field_name => Some(#field_idx)
+                        #unit{ .. } if #ref_name == #field_name => Some(#reflect_idx)
                     });
                     enum_name_at.push(quote! {
-                        #unit{ .. } if #ref_index == #field_idx => Some(#field_name)
+                        #unit{ .. } if #ref_index == #reflect_idx => Some(#field_name)
                     });
 
                     let field_ty = &field.data.ty;
-                    field_info.push(quote! {
-                        #bevy_reflect_path::NamedField::new::<#field_ty, _>(#field_name)
-                    });
-
-                    field_idx += 1;
-                }
-
-                let field_len = field_idx;
-                enum_field_len.push(quote! {
-                    #unit{..} => #field_len
-                });
-                enum_variant_name.push(quote! {
-                    #unit{..} => #name
-                });
-                enum_variant_type.push(quote! {
-                    #unit{..} => #bevy_reflect_path::VariantType::Struct
-                });
-                variant_info.push(quote! {
-                    #bevy_reflect_path::VariantInfo::Struct(
-                        #bevy_reflect_path::StructVariantInfo::new(#name, &[
-                            #(#field_info),*
-                        ])
-                    )
+                    quote! { #bevy_reflect_path::NamedField::new::<#field_ty, _>(#field_name) }
                 });
+                ("Struct", field_len, quote!(#name, &[ #(#field_info),* ]))
             }
-        }
+        };
+        let variant = Ident::new(variant_type, Span::call_site());
+        let info_type = format_ident!("{}VariantInfo", variant_type);
+        variant_info.push(quote! {
+            #bevy_reflect_path::VariantInfo::#variant(
+                #bevy_reflect_path::#info_type::new(#field_info)
```

Comment:
Imo this should be returned from the match arms. I find it more readable to see:

```rust
let field_info = quote! {
  #bevy_reflect_path::StructVariantInfo::new(#name, &[ #(#field_info),* ])
};

("Struct", field_len, field_info)
```

Since it gives the third field in that tuple context compared to:

```rust
("Struct", field_len, quote!(#name, &[ #(#field_info),* ]))
```

And this would remove the need for `format_ident!`.

Thoughts on this?

Reply:
I understand your concern. But I'm not a fan. N times the code means N times the chances for an overlooked mistake and N times the effort to fix, update or replace it. But here N = 3, it's not _horrible_. However, I've already been bitten many times on N = 3, so I'm very aggressive on DRY.

But to be fair, I don't like the code as it is now either. there is a lot of intertwined data dependencies which are not easy to spot at a glance. I'll try to strike a compromise.

Reply:
No need for a compromise. I meant to put this behind a "Nit", so you'd only need to change it if you also agreed 🙂